File: Data Structures & Algorithms/count-paths/submission-1.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        int M = m;
        int N = n;
        vector<vector<int>> dp(M+1,vector<int> (N+1,0));
        dp[M-1][N-1] = 1;
        for(int i = M-1; i>= 0; i--){
            for(int j = N-1; j>= 0; j--){
                dp[i][j]+= dp[i+1][j]+dp[i][j+1];
            }
        }
        return dp[0][0];
    }
};

File: Data Structures & Algorithms/burst-balloons/submission-1.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        n = len(nums)

        dp = [[-1] * n for _ in range(n)]

        def f(i, j):
            if i + 1 == j:
                return 0

            if dp[i][j] != -1:
                return dp[i][j]

            ans = 0

            for ind in range(i + 1, j):
                cost = (
                    nums[i] * nums[ind] * nums[j]
                    + f(i, ind)
                    + f(ind, j)
                )

                ans = max(ans, cost)

            dp[i][j] = ans
            return ans

        return f(0, n - 1)